src/files/fileset.rs

```diff
@@ -204,333 +286,101 @@ impl LoadedMod {
-                    entry.path().to_path_buf(),
-                    Vec::new(),
-                );
-                self.scan(dlc.root(), dlc.kind()).map_err(|e| FilesError::VanillaUnreadable {
-                    path: dlc.root().to_path_buf(),
-                    source: e,
-                })?;
-                self.loaded_dlcs.push(dlc);
-                add_loaded_dlc_root(label);
-            }
-        }
-        Ok(())
-    }
-
-    pub fn finalize(&mut self) {
-        // This sorts by pathname but where pathnames are equal it places `Mod` entries after `Vanilla` entries
-        // and `LoadedMod` entries between them in order
-        self.files.sort();
-
-        // When there are identical paths, only keep the last entry of them.
-        for entry in self.files.drain(..) {
-            if let Some(prev) = self.ordered_files.last_mut() {
-                if entry.path == prev.path {
-                    *prev = entry;
-                } else {
-                    self.ordered_files.push(entry);
-                }
-            } else {
-                self.ordered_files.push(entry);
-            }
-        }
-
-        for entry in &mut self.ordered_files {
-            let token = Token::new(&entry.filename().to_string_lossy(), (&*entry).into());
-            self.filename_tokens.push(token);
-            entry.store_in_pathtable();
-            self.filenames.insert(entry.path.clone());
-        }
-    }
-
-    pub fn get_files_under<'a>(&'a self, subpath: &'a Path) -> &'a [FileEntry] {
+    pub fn get_files_under<'a>(&'a self, subpath: &'a Path) -> &'a [Arc<FileEntry>] {
         let start = self.ordered_files.partition_point(|entry| entry.path < subpath);
         let end = start
             + self.ordered_files[start..].partition_point(|entry| entry.path.starts_with(subpath));
         &self.ordered_files[start..end]
     }
 
-    pub fn filter_map_under<F, T>(&self, subpath: &Path, f: F) -> Vec<T>
+    pub fn filter_map_under<'a, F, T>(&'a self, subpath: &'a Path, f: F) -> Vec<T>
     where
-        F: Fn(&FileEntry) -> Option<T> + Sync + Send,
+        F: Fn(&'a Arc<FileEntry>) -> Option<T> + Sync + Send,
         T: Send,
     {
         self.get_files_under(subpath).par_iter().filter_map(f).collect()
     }
 
     pub fn handle<T: Send, H: FileHandler<T>>(&self, handler: &mut H, parser: &ParserMemory) {
-        if let Some(config) = &self.config {
-            handler.config(config);
-        }
+        handler.config(&self.config);
         let subpath = handler.subpath();
         let entries = self.filter_map_under(&subpath, |entry| {
-            handler.load_file(entry, parser).map(|loaded| (entry.clone(), loaded))
+            handler.load_file(entry, parser).map(|loaded| (entry, loaded))
         });
         for (entry, loaded) in entries {
-            handler.handle_file(&entry, loaded);
+            handler.handle_file(entry, loaded);
         }
         handler.finalize();
     }
 
+    /// # Panics
```

Comment:
No need for this. The unwrap here is part of rust's "lock poisoning" feature. If a thread panics while holding a mutex, the mutex is marked as tainted. Unwrapping it ensures that all other threads will also panic if they try to take that lock, in order to ensure they don't encounter inconsistent data.

So it's not a real panic, it's a panic that's part of unrolling another panic.